pkg/machines/components/vm/vmExpandedContent.jsx

```diff
@@ -27,40 +27,65 @@ import VmNetworkTab from '../vmnetworktab.jsx';
 import Consoles from '../consoles.jsx';
 import VmOverviewTab from '../vmOverviewTabLibvirt.jsx';
 import VmUsageTab from './vmUsageTab.jsx';
+import VmSnapshotsTab from '../vmSnapshotsTab.jsx';
 import { ListingPanel } from 'cockpit-components-listing-panel.jsx';
 
 const _ = cockpit.gettext;
 
 /** One VM in the list (a row)
  */
-export const VmExpandedContent = ({
-    vm, vms, config, libvirtVersion, hostDevices, storagePools,
-    onUsageStartPolling, onUsageStopPolling, dispatch, networks, interfaces, nodeDevices, resourceHasError, onAddErrorNotification
-}) => {
-    const overviewTabName = (<a href="#" id={`${vmId(vm.name)}-overview`}>{_("Overview")}</a>);
-    const usageTabName = (<a href="#" id={`${vmId(vm.name)}-usage`}>{_("Usage")}</a>);
-    const disksTabName = (<a href="#" id={`${vmId(vm.name)}-disks`}>{_("Disks")}</a>);
-    const networkTabName = (<a href="#" id={`${vmId(vm.name)}-networks`}>{_("Network Interfaces")}</a>);
-    const consolesTabName = (<a href="#" id={`${vmId(vm.name)}-consoles`}>{_("Consoles")}</a>);
+export class VmExpandedContent extends React.Component {
+    constructor(props) {
+        super(props);
 
-    const tabRenderers = [
-        { name: overviewTabName, id: cockpit.format("$0-overview", vmId(vm.name)), renderer: VmOverviewTab, data: { vm, config, dispatch, nodeDevices, libvirtVersion } },
-        { name: usageTabName, id: cockpit.format("$0-usage", vmId(vm.name)), renderer: VmUsageTab, data: { vm, onUsageStartPolling, onUsageStopPolling }, presence: 'onlyActive' },
-        { name: disksTabName, id: cockpit.format("$0-disks", vmId(vm.name)), renderer: VmDisksTab, data: { vm, vms, config, storagePools, onUsageStartPolling, onUsageStopPolling, dispatch, onAddErrorNotification }, presence: 'onlyActive' },
-        { name: networkTabName, id: cockpit.format("$0-networks", vmId(vm.name)), renderer: VmNetworkTab, presence: 'onlyActive', data: { vm, dispatch, config, hostDevices, interfaces, networks, nodeDevices, onAddErrorNotification } },
-        { name: consolesTabName, id: cockpit.format("$0-consoles", vmId(vm.name)), renderer: Consoles, data: { vm, config, dispatch, onAddErrorNotification } },
-    ];
+        this.state = {
+            snapshotsAPIsAvailable: false,
+        };
+    }
 
-    let initiallyActiveTab = null;
-    if (vm.ui && vm.ui.initiallyOpenedConsoleTab) {
-        initiallyActiveTab = tabRenderers.map((o) => o.name).indexOf(consolesTabName);
+    componentDidMount() {
+        cockpit.spawn(["gdbus", "introspect", "--session", "--dest", "org.libvirt", "--object-path", "/org/libvirt/QEMU"])
```

Comment:
Please use cockpit.dbus() for that. We don't depend on gdbus, and calling a program is much  more expensive, too.

Also, this kind of "look before you leap" approach is brittle, as you still need to handle errors anyway. It's usually much more robust to just try and call the GetSnapshots() method, and handle the error case of "no such method".

Reply:
Sure I can do that. We keep snapshots list in redux store, so no dbus calls are done from this piece of code. But I can put some random snapshot dbus call here, I agree it's still better than calling gdbus thru command line.